x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/SparseVectorQueryBuilder.java

```diff
@@ -0,0 +1,367 @@
+            ? WeightedTokensUtils.queryBuilderWithPrunedTokens(fieldName, tokenPruningConfig, queryVectors, ft, context)
+            : WeightedTokensUtils.queryBuilderWithAllTokens(queryVectors, ft, context);
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+        if (queryVectors != null) {
+            return this;
+        } else if (weightedTokensSupplier != null) {
+            TextExpansionResults textExpansionResults = weightedTokensSupplier.get();
+            if (textExpansionResults == null) {
+                return this; // No results yet
+            }
+
+            return new SparseVectorQueryBuilder(
+                fieldName,
+                textExpansionResults.getWeightedTokens(),
+                null,
+                null,
+                shouldPruneTokens,
+                tokenPruningConfig
+            );
+        }
+
+        // TODO move this to xpack core and use inference APIs
+        CoordinatedInferenceAction.Request inferRequest = CoordinatedInferenceAction.Request.forTextInput(
+            inferenceId,
+            List.of(query),
+            TextExpansionConfigUpdate.EMPTY_UPDATE,
+            false,
+            InferModelAction.Request.DEFAULT_TIMEOUT_FOR_API
+        );
+        inferRequest.setHighPriority(true);
+        inferRequest.setPrefixType(TrainedModelPrefixStrings.PrefixType.SEARCH);
+
+        SetOnce<TextExpansionResults> textExpansionResultsSupplier = new SetOnce<>();
+        queryRewriteContext.registerAsyncAction(
+            (client, listener) -> executeAsyncWithOrigin(
+                client,
+                ML_ORIGIN,
+                CoordinatedInferenceAction.INSTANCE,
+                inferRequest,
+                ActionListener.wrap(inferenceResponse -> {
+
+                    List<InferenceResults> inferenceResults = inferenceResponse.getInferenceResults();
+                    if (inferenceResults.isEmpty()) {
+                        listener.onFailure(new IllegalStateException("inference response contain no results"));
+                        return;
+                    }
+                    if (inferenceResults.size() > 1) {
+                        listener.onFailure(new IllegalStateException("inference response should contain only one result"));
+                        return;
+                    }
+
+                    if (inferenceResults.get(0) instanceof TextExpansionResults textExpansionResults) {
+                        textExpansionResultsSupplier.set(textExpansionResults);
+                        listener.onResponse(null);
+                    } else if (inferenceResponse.getInferenceResults().get(0) instanceof WarningInferenceResults warning) {
+                        listener.onFailure(new IllegalStateException(warning.getWarning()));
+                    } else {
+                        listener.onFailure(
+                            new IllegalStateException(
```

Comment:
In this case we will return a 500 result. I think this is actually correct, because if we're throwing here there's a problem running inference.

Reply:
But the direct cause is that the user provided an inference ID that uses the wrong model type. So doesn't that make it user error?

Reply:
If you have a better suggestion of an Exception that should be thrown I'm happy to hear it, but otherwise I think this is OK.

Reply:
I suggest we throw `IllegalArgumentException` here, which will result in a 400 response to the user. This seems more appropriate to me since this case results from the user providing an inference ID which uses the wrong model (i.e. user error)

Reply:
I agree with Mike - it's a user error from my perspective, nothing went wrong calculating the inference.